model/converter/thrift/zipkin/to_domain.go

```diff
@@ -118,17 +131,48 @@ func (td toDomain) transformSpan(zSpan *zipkincore.Span) *model.Span {
 	if zSpan.ParentID != nil {
 		parentID = *zSpan.ParentID
 	}
-	return &model.Span{
+
+	flags := td.getFlags(zSpan)
+	result := []*model.Span{{
 		TraceID:       model.TraceID{High: uint64(traceIDHigh), Low: uint64(zSpan.TraceID)},
 		SpanID:        model.SpanID(zSpan.ID),
 		OperationName: zSpan.Name,
 		ParentSpanID:  model.SpanID(parentID),
-		Flags:         td.getFlags(zSpan),
+		Flags:         flags,
 		StartTime:     model.EpochMicrosecondsAsTime(uint64(zSpan.GetTimestamp())),
 		Duration:      model.MicrosecondsAsDuration(uint64(zSpan.GetDuration())),
 		Tags:          tags,
 		Logs:          td.getLogs(zSpan.Annotations),
+	}}
+
+	cs := td.findAnnotation(zSpan, zipkincore.CLIENT_SEND)
+	sr := td.findAnnotation(zSpan, zipkincore.SERVER_RECV)
+	if cs != nil && sr != nil {
+		// if the span is client and server we split it into two separate spans
+		s := &model.Span{
+			TraceID:       model.TraceID{High: uint64(traceIDHigh), Low: uint64(zSpan.TraceID)},
+			SpanID:        model.SpanID(zSpan.ID),
+			OperationName: zSpan.Name,
+			ParentSpanID:  model.SpanID(parentID),
+			Flags:         flags,
+		}
+		// if the first span is a client span we create server span and vice-versa.
+		if result[0].IsRPCClient() {
+			s.Tags = []model.KeyValue{model.String(ext.SpanKindRPCServer.Key, string(ext.SpanKindRPCServer.Value.(ext.SpanKindEnum)))}
```

Comment:
I think this is a lot simpler: `model.String(string(ext.SpanKind), string(ext.SpanKindRPCServerEnum))`

Reply:
+1 changed